### PR TITLE
Adds `StreamingDataLoader` class following the OLMo-core definition, and changes `grpo_fast.py` to use it.

### PR DESCRIPTION
**Pull request by finbarrtimbers**
Runs: 

1. Single GPU GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBNHWWHY6AT4WPMH1FXCNGWK?taskId=01KBNHWWJ1VAVH5GDBBAQC4YHW&jobId=01KBNHWWQAQK0Q5NWW0X78X1C8)
2. Single GPU GRPO with tools: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBNS32JVHNCK85PGVD9R49AT?taskId=01KBNS32K1ZGY6K66PK16WZGGG&jobId=01KBNS32PV9ASX2FKCZH6D1CD4)
3. Multi-node GRPO: [Beaker](https://beaker.allen.ai/orgs/ai2/workspaces/open-instruct-dev/work/01KBQPX2WMCRD8BQCPH7VAR477?taskId=01KBQPX2WQSX3QFPZ5MP565FN2&jobId=01KBQPX35PKQAWNDSXR00Y835G)

---

> [!NOTE]
> Adds a StreamingDataLoader with Ray-based DataPreparationActor and refactors GRPO to use streaming/vLLM configs, updating benchmarks, verifiers, utils, ActorManager, and tests.
> 
> - **Core Data Pipeline**:
>   - **New `StreamingDataLoader`** and configs (`StreamingDataLoaderConfig`, `VLLMConfig`) with Ray `DataPreparationActor` for centralized prompt generation, accumulation, packing, and per-rank batch delivery.
>   - Shared helpers: `accumulate_inference_batches`, `prepare_collated_data_for_workers`, `pad_sequences_for_world_size`, and `HFDataLoader`.
> - **GRPO (`grpo_fast.py`)**:
>   - Refactors to consume `StreamingDataLoader`/`DataPreparationActor`; removes local packing thread; training step pulls pre-packaged batches.
>   - Introduces `streaming_config`/`vllm_config` throughout (dataset/generation/reward/tool args migrated); updates checkpointing to include dataloader and actor state.
>   - Reworks vLLM engine setup, weight sync, metrics aggregation, and evaluation to use new pipeline.
> - **Actor Manager**:
>   - Accepts `streaming_config` and `vllm_config`; dashboard stats now compute batch size and concurrency from configs.
> - **Benchmarks**:
>   - `benchmark_generators.py` updated to use streaming/vLLM configs for dataset load, engine setup, sampling params, and CSV outputs.
> - **Verifiers/Utils**:
>   - `VerifierConfig.from_args` supports multiple sources; verifiers built with streaming config; set default for max-length verifier.
>   - `ModelDims.from_hf_config` and `MetricsTracker` tweaks; minor utility cleanups.
> - **Tests**:
>   - Update existing tests to use new data loader APIs; add GPU integration tests for `StreamingDataLoader` with/without tools.
> - **Misc**:
>   - Adds run instructions file for CI/manual runs.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 72540fac20fef0599309d05ecc5de9c6976f1064. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/data_loader.py [hunk 1]
```diff
@@ -127,3 +167,844 @@ def get_mock_batch(self) -> dict[str, Any]:
...
+        for i in range(num_initial_prompts):
+            add_prompt_to_generator(
+                next(self.iter_dataloader),
+                self.iter_dataloader._epoch,
+                self.param_prompt_Q,
+                self.generation_config,
+                is_eval=False,
+            )
+            if (i + 1) % 10 == 0 or i == num_initial_prompts - 1:
+                logger.info(f"[DataPreparationActor] Pushed {i + 1}/{num_initial_prompts} initial prompts")
```
Comment (hamishivi) [inline open_instruct/data_loader.py]: Maybe we could make this even lower log level, like debug? Feels not super useful unless you're literally debugging code.

File: open_instruct/data_loader.py [hunk 2]
```diff
@@ -127,3 +167,844 @@ def get_mock_batch(self) -> dict[str, Any]:
...
+                self.inference_results_Q,
+                self.generation_config,
+                num_prompts=self.global_batch_size,
+                model_dims=self.model_dims,
+                tokenizer=self.tokenizer,
+                dataset=self.dataset,
+                actor_manager=self.actor_manager,
+                active_sampling=self.config.active_sampling,
+                filter_zero_std_samples=self.config.filter_zero_std_samples,
+                replenish_prompts=True,
```
Comment (hamishivi) [inline open_instruct/data_loader.py:912]: Is there a reason to have replenish prompts as an argument if it's always true?

File: open_instruct/grpo_fast.py [hunk 3]
```diff
@@ -666,6 +516,36 @@ def to_device_inplace(tensors_list: list[torch.Tensor], device: torch.device):
 
 @ray.remote(num_gpus=1)
 class PolicyTrainerRayProcess(RayProcess):
+    def __init__(
```
[the remark this answers is not in the document]
Comment (hamishivi) [reply open_instruct/grpo_fast.py:343]: +1 on this. We should be able to save traces, this is really useful.

File: open_instruct/data_loader.py [hunk 4]
```diff
@@ -127,3 +167,844 @@ def get_mock_batch(self) -> dict[str, Any]:
             The first item from the dataset.
         """
         return self.dataset[0]
+
+
+@dataclass
+class StreamingDataLoaderConfig:
```
Comment (hamishivi) [inline open_instruct/data_loader.py:191]: The split of args b/w StreamingDataLoaderConfig and the original grpo_fast script is a little confusing to me. Number of samples per prompt is in here, but number unique prompts per batch is in grpo_Fast? Likewise, training stuff around how many mini batches etc. we take is in the grpo_Fast code, while async_steps is in here.

I don't have a great answer here, and to an outside user it doesn't matter, but it would be good I think to try to split up config handling a bit more cohesively...?

File: open_instruct/data_loader.py (hunk 1 above)
Comment (finbarrtimbers) [reply open_instruct/data_loader.py]: Removed!

File: open_instruct/data_loader.py (hunk 2 above)
Comment (finbarrtimbers) [reply open_instruct/data_loader.py:912]: We set it to be False in `grpo_fast.py`, when accumulating eval prompts!

File: open_instruct/data_loader.py (hunk 4 above)
Comment (finbarrtimbers) [reply open_instruct/data_loader.py:191]: I took another pass at this! Please let me know what you think.

**Review by hamishivi — APPROVED** [review]
